### ISSUE TITLE
No guaranteed way to register routes in order

### ISSUE THREAD

**Issue by corytheboyd** [jfairbank/redux-saga-router#16]
It would be nice to be able to pass in routes as an Array instead of an Object so you know the order in which routes are registered with ruta3 is stable. Either that, or allow passing of a route matching object through options. Thoughts?

**Comment by jfairbank** [comment]
Hi @corytheboyd!

Are you running into an issue with the object syntax currently? Just trying to understand your use case for needing ordered routes.

**Comment by corytheboyd** [comment]
Hey @jfairbank!

We haven't run into any issues registering routes with the object syntax, it's just that the language spec does not guarantee the order of properties for objects afaik. We're actually defining our routes in an Array and converting that to an object per the `redux-saga-router` API. We're working with something like this at the moment (simplified version):

```javascript
// src/js/routes.js

import { fork } from 'redux-saga/effects';

class Route {
  constructor(options) {
    const {
      pattern,
      handler,
    } = options;

    if (!pattern) {
      throw new Error('Route must specify a pattern');
    }

    this.pattern = pattern;
    this.handler = handler;
  }
}

const ROUTES = [
  new Route({ pattern: '/foo', handler: function* handlerSaga() {} }),
  new Route({ pattern: '/foo/exactPath', handler: function* handlerSaga() {} }),
  new Route({ pattern: '/foo/:ambiguousPath', handler: function* handlerSaga() {} }),
]

// Converting our Array structure to an Object
const routesObject = ROUTES.reduce((acc, route) => {
  acc[route.pattern] = route.handler;
  return acc;
}, {});

// root saga later forked along with other always-running sagas
export default function* root() {
  yield fork(router, history, routesObject, options);
}
```

This completely works as is, including order preservation, but that conversion step feels a bit unnecessary, given that an Array is inherently ordered. Perhaps the `buildRouteMatcher` function could just be modified to check the type of the `routes` parameter to change the iteration strategy. https://github.com/jfairbank/redux-saga-router/blob/master/src/buildRouteMatcher.js#L6

Quick pseudo-code for an enhanced `buildRouteMatcher`: 
```javascript
export default function buildRouteMatcher(routes) {
  const routeMatcher = ruta3();

  // No actual need for the do block, just for example
  const routesList = do {
    if (isObject(routes)) { // Where isObject is a suitable function for type checking
      Object.keys(routes);
    } else if (Array.isArray(routes)) {
      routes
    } else {
      throw new Error('routes definition must be an Object or an Array');
    }
  };

  routesList.forEach((route) => {
    routeMatcher.addRoute(route, routes[route]);
  });

  return routeMatcher;
}
```

**Comment by eL-HaXo** [comment]
I'd also like to see this issue implemented. I'm trying to define a saga that is run in the event that no other routes are matched (a 404 saga). Using the pattern `*` or `/*` will run the saga first no matter where it is positioned in the route object (I am doing a similar transformation as OP). If I use the option `{ matchAll: false }`, none of the other sagas ever run.

**Comment by visusnet** [comment]
@jfairbank This is fixed with 2.2.0, isn‘t it?

**Comment by jfairbank** [comment]
@visusnet Yes. Closing this.
